kotlin/workflow-runtime/src/main/java/com/squareup/workflow/debugging/WorkflowUpdateDebugInfo.kt

```diff
@@ -0,0 +1,106 @@
+import com.squareup.workflow.debugging.WorkflowUpdateDebugInfo.Kind.ChildDidUpdate
+import com.squareup.workflow.debugging.WorkflowUpdateDebugInfo.Kind.DidUpdate
+import com.squareup.workflow.debugging.WorkflowUpdateDebugInfo.Source
+import com.squareup.workflow.debugging.WorkflowUpdateDebugInfo.Source.External
+import com.squareup.workflow.debugging.WorkflowUpdateDebugInfo.Source.Subtree
+import com.squareup.workflow.debugging.WorkflowUpdateDebugInfo.Source.Worker
+import com.squareup.workflow.internal.WorkflowId
+
+/**
+ * A description of a workflow update triggered by a [Source] (worker, event, etc).
+ *
+ * This is a simple linked list that represents a traversal down the workflow tree that starts at
+ * the root and indicates, for each workflow, if its child output something that it handled
+ * (see [Kind]), or if it was just a parent of a workflow that didn't output anything.
+ *
+ * When a workflow handles an update, the type of update is indicated by [Source].
+ */
+data class WorkflowUpdateDebugInfo(
+  val workflowType: String,
+  val kind: Kind
+) {
+  constructor(
+    workflowId: WorkflowId<*, *, *>,
+    kind: Kind
+  ) : this(workflowId.typeDebugString, kind)
+
+  /**
+   * A sealed class that indicates whether a workflow actually executed a `WorkflowAction`, or
+   * was just the ancestor of a workflow that did.
+   *
+   * Contains two subclasses, see their documentation for details:
+   * - [DidUpdate]
+   * - [ChildDidUpdate]
+   */
+  sealed class Kind {
+    // TODO more kotliny names
+    /**
+     * Indicates that this workflow actually executed a `WorkflowAction`.
+     * [DidUpdate.source] is a [Source] that indicates if the action was triggered by:
+     * - An event ([Source.External])
+     * - A worker ([Source.Worker])
+     * - A child workflow emitting an output ([Source.Subtree])
+     */
+    data class DidUpdate(val source: Source) : Kind()
+
+    /**
+     * Indicates that one of this workflow's descendants executed a
+     * `WorkflowAction`, but none of its immediate children emitted an output, so this workflow
+     * didn't get directly notified about it.
+     */
+    data class ChildDidUpdate(val childInfo: WorkflowUpdateDebugInfo) : Kind()
+  }
+
+  /**
+   * A sealed class that indicates what triggered the update.
+   *
+   * Contains three subclasses, see their documentation for details:
+   * - [External]
+   * - [Worker]
+   * - [Subtree]
+   */
+  sealed class Source {
+    /**
+     * Indicates that the update was triggered by an event being received by a
+     * [Sink][com.squareup.workflow.Sink] (see
+     * [makeActionSink][com.squareup.workflow.RenderContext.makeActionSink]).
+     */
+    object External : Source()
+
+    /**
+     * Indicates that the update was triggered by an output emitted from a
+     * [Worker][com.squareup.workflow.Worker] being run by this workflow.
+     */
+    data class Worker(val key: String) : Source()
```

Comment:
Should this class (and `Subtree` below) also include the actual result of the update (i.e. the output value that was emitted)? Seems like that could be helpful for debugging, but again I'm not sure why Swift was reluctant to include raw values in these types.

Reply:
Is Swift always doing this eager string rendering in production?

Reply:
I believe so.